sdks/java/extensions/euphoria/src/main/java/org/apache/beam/sdk/extensions/euphoria/core/translate/provider/GenericTranslatorProvider.java

```diff
@@ -0,0 +1,246 @@
+      possibleTranslators.add(TranslationDescriptor.of(predicate, operatorTranslator));
+      return this;
+    }
+
+    public GenericTranslatorProvider build() {
+      return new GenericTranslatorProvider(possibleTranslators);
+    }
+  }
+
+  /**
+   * Container for optional {@link Predicate user defined predicate}, optional {@link Class} of na
+   * operator to be translated and {@link OperatorTranslator} itself. The predicate and operator
+   * class defines an optional checks. At least one of them have to be present.
+   *
+   * <p>The {@link OperatorTranslator} is allowed to translate an operator iff it pass all the
+   * checks and {@link OperatorTranslator#canTranslate(Operator) can translate} given operator. That
+   * allows users to write translators specific for any operator.
+   *
+   * @param <OperatorT> the type of the euphoria operator
+   */
+  private static class TranslationDescriptor<OperatorT extends Operator<?>> {
+
+    /** Class of an {@link Operator} given {@link #translator} can be applied on. */
+    private final Optional<Class<OperatorT>> operatorClass;
+
+    /**
+     * User specified predicate, which determines whenever given {@link #translator} can be used to
+     * translate an operator.
+     */
+    private final Optional<Predicate<OperatorT>> userDefinedPredicate;
+
+    private final OperatorTranslator<?, ?, OperatorT> translator;
+
+    private TranslationDescriptor(
+        Optional<Class<OperatorT>> operatorClass,
+        Optional<Predicate<OperatorT>> userDefinedPredicate,
+        OperatorTranslator<?, ?, ? extends OperatorT> translator) {
+      Preconditions.checkState(
+          operatorClass.isPresent() || userDefinedPredicate.isPresent(),
+          "At least user defined predicate or class of an operator have to be given.");
+
+      @SuppressWarnings("unchecked")
+      OperatorTranslator<?, ?, OperatorT> castedTranslator =
+          (OperatorTranslator<?, ?, OperatorT>) translator;
+
+      this.operatorClass = operatorClass;
+      this.userDefinedPredicate = userDefinedPredicate;
+      this.translator = castedTranslator;
+    }
+
+    static <OperatorT extends Operator<?>> TranslationDescriptor<OperatorT> of(
+        Class<OperatorT> operatorClass,
+        Predicate<OperatorT> userDefinedPredicate,
+        OperatorTranslator<?, ?, ? extends OperatorT> translator) {
+      return new TranslationDescriptor<>(
+          Optional.of(requireNonNull(operatorClass)),
+          Optional.of(requireNonNull(userDefinedPredicate)),
+          requireNonNull(translator));
+    }
+
+    static <OperatorT extends Operator<?>> TranslationDescriptor<OperatorT> of(
+        Predicate<OperatorT> userDefinedPredicate,
+        OperatorTranslator<?, ?, ? extends OperatorT> translator) {
+      return new TranslationDescriptor<>(
+          Optional.empty(),
+          Optional.of(requireNonNull(userDefinedPredicate)),
```

Comment:
unnecessary null check - https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#of-T-

Reply:
fixed